### ISSUE TITLE
Hot Reload Hang on MacOS

### ISSUE THREAD

**Issue by nick-thompson** [nick-thompson/blueprint#79]
I'm finding that during a hot reload on MacOS, Xcode just hangs in a permanent loop and the project freezes.

Dropping the debugger in shows that we're stuck in the Duktape finalizer callstack searching through the `lambdaReleasePool`, but it seems like there are a million things to be finalized. It looks like there indeed might be a bug with the lambdaReleasePool, because it's huge, but I'm not sure that's exactly what's hanging up Xcode here.

@JoshMarler have you seen this one?

**Comment by JoshMarler** [comment]
Interesting, I have seen this previously when the "animate" prop was set on the canvas element. The clearing of the duk context and pool on a reload appeared to resolve the issues I was seeing but I'm not using animated draws. I'll be back at me desk shortly so I'll fire this up on Mac and take a look.

**Comment by nick-thompson** [comment]
Ahh right, you mentioned those fixes before. Yea, they sound like a good idea to me.

I just noticed that I can accelerate the lock up by resizing the window a bunch in gain plugin. Just making note of that here

**Comment by JoshMarler** [comment]
Yeah. I seem to remember it appeared to be mostly antagonised by the `animate` functionality on `CanvasView` but I'll have to take another look. Does seem there might be some little lurking thing between the release pool and duktape. I think the exponential growth of the pool existed from the first Canvas prototype but if I've introduced something mad apologies!

I'm just on separate thing at the moment but I will for sure take a look at this today. I'll let you know if I get anywhere with it.

**Comment by nick-thompson** [comment]
Yea, I think you're right.  I don't think it was something you introduced, I think the finalizer stuff I wrote is probably buggy and not always removing lambdas correctly. If you have time I would definitely appreciate the look! I'll try to put some time into it later this week as well

**Comment by JoshMarler** [comment]
Yeah no worries. I'll take a look today/tomorrow for sure.

**Comment by nick-thompson** [comment]
:D Yea, this might be a gnarly one. Thank you!

**Comment by JoshMarler** [comment]
Heh. Writing a little level meter component at the minute and getting freezes and terrible performance .... I have a sneaking suspicion this has just bitten me. Into the bug dive I go!

**Comment by nick-thompson** [comment]
😁 good luck! I haven't had a chance to dig in yet; hope it's not too gnarly

**Comment by JoshMarler** [comment]
Looking through this at the moment. At a first pass, it almost looks like there is a significant lag in the call the the finalizers. From Duktape docs:

"Finalizers are executed for unreachable objects detected by reference counting or mark-and-sweep. The finalizer may not execute immediately, however, not even when reference counting detects that the object became unreachable."

I wonder if it literally is taking this long for the finalizers to be called after garbage collection. Hence our filling up of the pool which does eventually get cleared, though it blocks the UI/Message thead.

So high level thoughts are, can we increase GC rate? Can we reduce the number of LambdaHelpers being created for repeated Canvas redraws? If this lag is unavoidable should the release pool live on some backkground thread which can be incrementally cleared without blocking the message loop? 

This comment is more me thinking aloud than anything at the moment..., will continue digging.

**Comment by JoshMarler** [comment]
Also worth noting: 

>If Duktape runs out of memory (despite emergency GC) when trying to call a finalizer, the call error is silently ignored and the finalizer will be skipped.

>When an object is finalized by mark-and-sweep but becomes unreachable before the next mark-and-sweep round has a change to detect the rescue, the object's finalizer will not be executed.

**Comment by JoshMarler** [comment]
Based on look under the debugger it does look like the rate at which we push lambdas into the pool when painting on a timer is far higher than the rate of GC. We push enough in there for the loops over the pool to significantly block the UI thread. 

A first experiment might be trying a different data-structure for the pool to increase the lookup times. unordered_set of pointers might be worth a try and preallocate some sensible amount of memory?

Failing that maybe we need a different scheme for the canvas onDraw property function invocations.

**Comment by JoshMarler** [comment]
So yeah, a quick hack switching to `std::unordered_set<LambdaHelper*>` resolves the freeze. Obviously this sort of approach would need some tidying. We could add a key to LambdaHelper using a static counter and use unordered_map with that key to avoid lifetime traps. 

I might think about this more though and workout if anything nicer can be done. That's a lot of small allocations and deallocations  flying around.

Any thoughts on this one?

**Comment by JoshMarler** [comment]
Might also be interesting to manually call duk_gc after invoking the lambdas here. 

https://github.com/svaarala/duktape/issues/1550

I wonder if the "light function" concept is also worth some investigation to avoid creating JS function objects in some cases which appear to require periodic garbage collection to destroy rather than being immediately released via ref counting. I need to read into this all in more detail ....

**Comment by JoshMarler** [comment]
Also interesting: https://github.com/svaarala/duktape/issues/1872

**Comment by JoshMarler** [comment]
OK after tidying a little using unordered_set performance isn't so bad...

**Comment by JoshMarler** [comment]
I'd be interested to look into light funcs for this still. `LamdaHelper::invokeFromDukContext` is static so can reasonably delete the LambdaHelper object once it's `callback` has been invoked. This could help move away from the finalizer and release pool approach. I don't think there are any major disadvantages to using light funcs in this context? What do you think @nick-thompson ? Is this a worthwhile experiment? 

I can drop over the fix for the current freeze in the meantime and look into this as a separate bit of work if needed.

**Comment by JoshMarler** [comment]
I'm wondering whether we can avoid pushing a new canvas context object to the duk stack for every single invocation and if we can instead store a canvas context object permanently in the global namespace or similar ...

**Comment by JoshMarler** [comment]
Ahh OK. The freeze is much improved in the gain example but there are still occasional lags ...

**Comment by JoshMarler** [comment]
Sorry I'm spamming you a bit but I've been profiling so I'll register findings here. 

. Rendering text via strokeText and fillText on the canvas object is currently a hotspot. Removing that made a fair amount of difference. I've got some text overlaid on this meter. I should probably have two components with one using a higher z-order, I'm not sure how z-order is intended to be controlled via component props at this stage? Is this still a thing to be implemented @nick-thompson ? Relative z orderings on views?

. The canvas `drawImage` function is also pretty costly, no great suprises there either. 

. Hotspot is various locations in `duk_pcall` inside `EcmascriptEngine::readVarFromDukStack` in `CanvasView::paint`. No great suprises there. Most time is spent in garbage collection functions, object property lookup (`duk_hobject_getptop`) and property setting (maybe from repeated calls to `bindCanvasContextProperties` ? I have an idea for that).

**Comment by JoshMarler** [comment]
OK ... I'm clearly also redrawing way too many things when I should just be drawing my meter ... some prop/state lifecycle thing happening I'm yet to catch/grasp.

**Comment by nick-thompson** [comment]
@JoshMarler you're the man, this is so awesome. I'll try to address everything in a few high level thoughts, to start:

1. Yep, I'm down for `unordered_set`, that's a good calll
2. Yes indeed! #5 
5. I'd be curious to see a before/after profile of how much we get from that. If it's really significant, I'm interested
6. I didn't know about these, but I'm very interested to have them in the docs or if we can find a way to integrate duktape better (#62), enabling these things by default
8. Definitely when looking into this, make sure you're using `shouldComponentUpdate` or the new hooks API to ensure components don't update when they don't need to. That's always a huge performance help.

High level I'm thinking basically that there are two types of functions that we push to javascript: persistent functions via `registerNativeMethod`, and the more transient functions like `rect()` on the `CanvasContext` that we prepare and push onto the stack for invoking `onDraw` in the `CanvasView`. In the first case, we obviously need to hold onto that lambda to keep the function implementation alive. I think that's a strong use case for the duktape finalizers. The latter case, I feel like we could probably find a way to push the CanvasContext onto the stack, call `onDraw`, and then immediately kill the lambdas that we had registered for that call. After `onDraw` that `CanvasContext` won't be on the stack anymore, and it's almost certainly the case that the last reference has been dropped after that call completes. We could manually invoke the GC right there, but I think that's overkill. I'm wondering if there's a way to distinguish that we're taking the "transient" path here, avoid ever pushing these functions into the lambdaRelasePool, and allow their destruction when the current C++ stack frame unwinds.

I need to think on this for a bit, but I feel like there's likely an elegant solution there...

Thanks a million for the detail here @JoshMarler, I'll write back with more thoughts soon

**Comment by JoshMarler** [comment]
@nick-thompson, 

No Problem! It's been massively valuable to dig into this. I found the causes for my wasted reconciles. Indeed, I reckon the `PureComponent` idea might be worth spattering around my code base to assist when others make similar mistakes etc. 

The idea around CanvasContext is interesting, I totally agree. I wondered about the whole "light function" thing and whether that is some we could somehow leverage but I'm still not sure yet. I did briefly experiment with having a single `CanvasContext` registered as a global property, it sort of worked with much odd behaviour, basically `ReactApplicationRoot` owned the instance and passed it into the `CanvasView`'s constructor. I don't think that's a particularly good idea though. 

If we could workout someway for things like canvas `onDraw` calls running on timers to put less stress on the GC I think that would be a HUGE win. The performance is actually looking really good around all of this now. 

I'll get the unordered_set fix over to you and deal with #5 as a start.

**Comment by JoshMarler** [comment]
@nick-thompson, PR with fix for the freeze here: https://github.com/nick-thompson/blueprint/pull/81

**Comment by nick-thompson** [comment]
Ahh yea sorry, I skipped right over the `z-order` question. So `z-order` is already sort of implemented based on literally the order of react children:

```
<View>
  <View />
  <View />
  <Modal /> // Move this to the top; the first child of the parent `View` will have the top-most z order
</View>
```

Implementing a proper `z-index` property implementation is on the TODO list, and should be quite straightforward: by default we continue with the default z-order behavior (based on the order of the children in the parent), but allow a `z-index` property (https://developer.mozilla.org/en-US/docs/Web/CSS/z-index) whose behavior is implemented by walking the parent's list of children and moving the child appropriately using https://docs.juce.com/master/classComponent.html#a1b0cca8b9a9fea673aefc84a42602bcd.

Definitely happy to have you tackle #5 and #4. Same PR is cool with me. For #5 I want to dig into the React HostConfig a little more to see if there really is no "commit" callback, as if to say "we've now applied all property updates, commit the update transaction." That certainly seems like the ideal time to do the layout. If not, I think we should use something like https://docs.juce.com/master/classAsyncUpdater.html to coalesce the updates. For example, if I set `width: 100`, we enqueue the layout callback via `triggerAsyncUpdate`. But then maybe another property change for `height: 100` comes through and we call that same `triggerAsyncUpdate`, and if this is all synchronous then the previous pending update will still be sitting there and we get automatic coalescence. Mixing this into the updates from #4 and only triggering the async update if the set of properties changed involved layout-related properties seems like the big winner.

I'll take a look at the PR now, thanks so much for digging into it. I want to spend some time as well playing with Duktape, lightfuncs, and exploring this GC thing and "transient" function stack objects. I'll write back if I find anything
